packages/core/src/components/AutocompleteField/AutocompleteField.jsx

```diff
@@ -0,0 +1,198 @@
+import Button from '../Button/Button';
+import Downshift from 'downshift';
+import PropTypes from 'prop-types';
+import React from 'react';
+import TextField from '../TextField/TextField';
+import uniqueId from 'lodash.uniqueid';
+
+/**
+ * Determine if a React component is a TextField
+ * @param {React.Node} child - a React component
+ * @return {Boolean} Is this a TextField component?
+ */
+function isTextField(child) {
+  return child != null && child.type === TextField;
+}
+
+/**
+ * An `AutocompleteField` component renders a typeahead input field, form label, and a clear input link.
+ */
+export class AutocompleteField extends React.PureComponent {
+  constructor(props) {
+    super(props);
+
+    this.id = uniqueId('autocomplete_');
+  }
+
+  renderChildren(getInputProps) {
+    // Extend props on the TextField, by passing them through
+    // Downshift's `getInputProps` method
+    return React.Children.map(this.props.children, child => {
+      const { disabled } = this.props;
+
+      if (isTextField(child)) {
+        return React.cloneElement(
+          child,
+          getInputProps({
+            className: disabled ? 'ds-c-autocomplete__inner--disabled' : '',
+            disabled: disabled,
+            fieldClassName: 'ds-c-autocomplete__input',
+            id: this.id,
+            inputId: this.id,
+            labelClassName: 'ds-u-margin-top--0 ds-c-autocomplete__label'
+          })
+        );
+      }
+
+      return child;
+    });
+  }
+
+  render() {
+    const {
+      clearAriaLabel,
+      clearInputText,
+      constrainedList,
+      constrainedListText,
+      items,
+      itemToString,
+      onChange,
+      onStateChange
+    } = this.props;
+
+    return (
+      <Downshift
+        itemToString={itemToString}
+        onChange={onChange}
+        onStateChange={onStateChange}
+        render={({
+          clearSelection,
+          getInputProps,
+          getItemProps,
+          highlightedIndex,
+          isOpen,
+          inputValue
+        }) => (
+          <div className="ds-u-clearfix ds-c-autocomplete">
+            {this.renderChildren(getInputProps)}
+
+            {isOpen ? (
+              <div className="ds-u-border--1 ds-u-padding--1 ds-c-autocomplete__list">
+                {constrainedList ? (
+                  <h5
+                    className="ds-u-margin--0 ds-u-padding--1"
+                    id="constrained-list-header"
+                  >
+                    {constrainedListText}
+                  </h5>
+                ) : null}
+
+                <ul
+                  aria-labelledby={
+                    constrainedList ? 'constrained-list-header' : null
+                  }
+                  className="ds-c-list--bare"
+                  role="listbox"
+                >
+                  {items
```

Comment:
This component currently renders an empty list if `items` is empty. Can we display a (customizable) "No results" message? Also, how would this work if you have to query an API..is there a loading state?

![image](https://user-images.githubusercontent.com/371943/34017113-5aa1e6a6-e0f2-11e7-8dfe-49ab3e049b02.png)

![image](https://user-images.githubusercontent.com/371943/34017117-5d8e3180-e0f2-11e7-8134-4e8b866eaa26.png)